### PR TITLE
Update docs to include disable_reflected_list switch

### PR DESCRIPTION
**Pull request by esc**
This add documentation for the `disable_reflected_list` switch to the reference
manual and the dperecation documentation.

Questions:

a) will this suffice or should we add some more info to deprecation docs.
b) do we want to switch from PendingDeprecationWarning to DeprecationWarning
already?

### CODE REVIEW

File: docs/source/reference/pysupported.rst [hunk 1]
```diff
@@ -552,6 +556,119 @@ Finally, here's an example of using a nested `List()`:
...
 
+.. _feature-disable-reflected-list:
+
+
+Disable Reflected List
+''''''''''''''''''''''
+
+As of Numba 0.47.0 a new switch, ``disable_reflected_list`` has been introduced
+to the ``numba.future`` module. Using this switch will disable the
+*reflected-list* fully across the Numba code-base. This means, that the
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]:
```suggestion
*reflected-list* fully across the Numba code-base. This means that the
```

File: docs/source/reference/pysupported.rst [hunk 2]
```diff
@@ -552,6 +556,119 @@ Finally, here's an example of using a nested `List()`:
...
+built-in constructors ``list()`` and ``[]`` will result in a *typed-list* Also,
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]:
```suggestion
built-in constructors ``list()`` and ``[]`` will result in a *typed-list*. Also,
```

File: docs/source/reference/pysupported.rst [hunk 3]
```diff
@@ -552,6 +556,119 @@ Finally, here's an example of using a nested `List()`:
...
+
+.. code-block:: python
+
+    from numba import njit
+    from numba.future import disable_reflected_list
+
+    @njit
+    def foo():
+        a = list()     # create an empty typed-list
+        b = []         # create another empty typed-list
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]: this won't compile, it has no type

File: docs/source/reference/pysupported.rst [hunk 4]
```diff
@@ -552,6 +556,119 @@ Finally, here's an example of using a nested `List()`:
...
+        c = [1, 2, 3]  # create typed-list containing 1, 2 and 3
+        return a, b, c # return all three
+
+
+Additionally, it is possible to hand Python lists into Numba compiled functions
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]:
```suggestion
Additionally, it is possible to pass Python lists into Numba compiled functions
```

File: docs/source/reference/pysupported.rst [hunk 5]
```diff
@@ -552,6 +556,119 @@ Finally, here's an example of using a nested `List()`:
...
+        acc = 0
+        # if lst was a Python list, Numba has converted it into an immutable
+        # typed-list
+        for i in lst:  
+            acc += i
+        return acc
+
+    total = foo(z)
+
+Immutability in this case, means that the contents of the *typed-list* can not
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]:
```suggestion
Immutability in this case, means that the contents of the *typed-list* cannot
```

File: docs/source/reference/pysupported.rst [hunk 6]
```diff
@@ -552,6 +556,119 @@ Finally, here's an example of using a nested `List()`:
...
+be modified, for example using methods such as `append`, `pop` or `clear`.
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]:
```suggestion
be modified, for example, by using methods such as `append`, `pop` or `clear`.
```

File: docs/source/reference/pysupported.rst [hunk 7]
```diff
@@ -552,6 +556,119 @@ Finally, here's an example of using a nested `List()`:
...
+        lst.append(3)
+
+    l = List()
+    l.append(1)
+    l.append(2)
+
+    append_three(l)
+    # l now contains 1, 2, 3
+
+An important advantage over the previous behaviour of the reflected list, is
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]:
```suggestion
An important advantage over the previous behaviour of the reflected list is
```

File: docs/source/reference/pysupported.rst [hunk 8]
```diff
@@ -427,7 +427,10 @@ list
     However, it was scheduled for deprecation from version 0.44.0 onwards due
     to its limitations. As of version 0.45.0 a new implementation, the
     so-called *typed-list* (see below), is available as an experimental
-    feature. For more information, please see: :ref:`deprecation`.
+    feature. As of version 0.47.0 a new switch, ``disable_reflected_list`` is
+    available from the ``numba.future`` module that can be used to opt-in to
+    fully disable the reflected list for all use-cases (see below).
+    For more information, please see: :ref:`deprecation`.
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]:
```suggestion
    For more information about the deprecation schedule, please see: :ref:`deprecation`.
```

File: docs/source/reference/pysupported.rst [hunk 9]
```diff
@@ -552,6 +556,119 @@ Finally, here's an example of using a nested `List()`:
    :dedent: 4
    :linenos:
 
+.. _feature-disable-reflected-list:
+
+
+Disable Reflected List
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]:
```suggestion
Disabling Reflected Lists
```
(and update section header markup in line below)

File: docs/source/reference/pysupported.rst [hunk 10]
```diff
@@ -552,6 +556,119 @@ Finally, here's an example of using a nested `List()`:
...
+''''''''''''''''''''''
+
+As of Numba 0.47.0 a new switch, ``disable_reflected_list`` has been introduced
+to the ``numba.future`` module. Using this switch will disable the
+*reflected-list* fully across the Numba code-base. This means, that the
+built-in constructors ``list()`` and ``[]`` will result in a *typed-list*. Also,
+this means that any Python list that is passed into a Numba compiled function
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]: suggestion: `This also means that...`

File: docs/source/reference/deprecation.rst [hunk 11]
```diff
@@ -124,6 +124,12 @@ are:
   formats are largely avoided.
 * Numba's ``typed.Dict`` will be able to use these containers as values.
 
+As of Numba 0.47.0 a feature switch, `disable_reflected_list` has been
```
Comment (stuartarchibald) [inline docs/source/reference/deprecation.rst]:
```suggestion
As of Numba 0.47.0 a feature switch, ``disable_reflected_list`` has been
```

**Comment by stuartarchibald** [comment]
This implements the actual deprecation of reflection: https://github.com/numba/numba/pull/4984

**Comment by seibert** [comment]
The deprecation schedule makes sense.  If we insert a short extra release, we can revise things.

File: docs/source/reference/pysupported.rst [hunk 12]
```diff
@@ -514,9 +517,10 @@ explicitly from the `numba.typed` module::
 
 
 .. note::
-    As the typed-list stabilizes it will fully replace the reflected-list and the
-    constructors `[]` and `list()` will create a typed-list instead of a
-    reflected one.
+    The typed-list is beginning to stabilize with Numba version 0.47.0. Using
```
[the remark this answers is not in the document]
Comment (esc) [reply docs/source/reference/pysupported.rst]: Depends, it is still missing a few dunder methods and might have a refcount bug in `pop`. However, it hasn't significantly caught fire or caused a cascade failure anywhere we know of, so it will probably be alright. And if push comes to shove we can always hotfix and release a patch version.

[the remark this answers is not in the document]
Comment (esc) [reply docs/source/reference/pysupported.rst]: So my answer is: yes we should consider it stable.

File: docs/source/reference/pysupported.rst (hunk 10 above)
Comment (esc) [reply docs/source/reference/pysupported.rst]: sure

File: docs/source/reference/pysupported.rst (hunk 3 above)
Comment (esc) [reply docs/source/reference/pysupported.rst]: indeed

**Comment by esc** [comment]
PR has been updated and all comment addressed.

File: docs/source/reference/pysupported.rst [hunk 13]
```diff
@@ -552,6 +557,122 @@ Finally, here's an example of using a nested `List()`:
...
 
+.. _feature-disable-reflected-list:
+
+
+Disabling Reflected Lists
+''''''''''''''''''''''''''
+
+As of Numba 0.47.0 a switch, ``disable_reflected_list``, has been introduced to
+the ``numba.future`` module. Using this switch will disable the
+*reflected-list* fully across the Numba code-base. This means that the built-in
```
Comment (sklam) [inline docs/source/reference/pysupported.rst]: > fully across the Numba code-base

To me, it sounds like it would affect every module that uses list. 

Maybe rephrase to:

`Referencing ``disable_reflected_list`` in a module will disable *reflected-list* fully in all Numba-compiled code inside that module.`

Comment (esc) [reply docs/source/reference/pysupported.rst]: done

File: docs/source/reference/pysupported.rst [hunk 14]
```diff
@@ -514,9 +518,10 @@ explicitly from the `numba.typed` module::
 
 
 .. note::
-    As the typed-list stabilizes it will fully replace the reflected-list and the
-    constructors `[]` and `list()` will create a typed-list instead of a
-    reflected one.
+    The typed-list is considered stable with Numba version 0.47.0. Using
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]:
```suggestion
    The typed-list is considered stable in Numba version 0.47.0. Using
```

**Comment by esc** [comment]
> Thanks for the fixes (1 typo I just noticed to sort still). I think the only remaining thing is to actually change the deprecation schedule with respect to [#4982 (review)](https://github.com/numba/numba/pull/4982#pullrequestreview-334025175), i.e.:
> 
> ```
> List:
> 
>     Deprecate 0.47
>     Remove 0.48
> 
> Set:
> 
>     Deprecate 0.48
>     Remove 0.49
> ```
> 
> thanks.

@stuartarchibald I believe I have updated it accordingly here: https://github.com/numba/numba/pull/4982/files#diff-289bdad66c3d6f45cc36485c6a8c4371R102-R112 or did I miss something?

**Comment by stuartarchibald** [comment]
> > Thanks for the fixes (1 typo I just noticed to sort still). I think the only remaining thing is to actually change the deprecation schedule with respect to [#4982 (review)](https://github.com/numba/numba/pull/4982#pullrequestreview-334025175), i.e.:
> > ```
> > List:
> > 
> >     Deprecate 0.47
> >     Remove 0.48
> > 
> > Set:
> > 
> >     Deprecate 0.48
> >     Remove 0.49
> > ```
> > 
> > 
> > thanks.
> 
> @stuartarchibald I believe I have updated it accordingly here: https://github.com/numba/numba/pull/4982/files#diff-289bdad66c3d6f45cc36485c6a8c4371R102-R112 or did I miss something?

ah yes, thanks, somehow missed that.

**Comment by esc** [comment]
🚢

**Comment by esc** [comment]
Closing this as #4883 has been abandoned.
